tensorflow/core/profiler/internal/gpu/rocm_tracer.h

```diff
@@ -0,0 +1,302 @@
+
+ private:
+  struct AnnotationMapImpl {
+    // The population/consumption of annotations might happen from multiple
+    // callback/activity api related threads.
+    absl::Mutex mutex;
+    // Annotation tends to be repetitive, use a hash_set to store the strings,
+    // an use the reference to the string in the map.
+    absl::node_hash_set<std::string> annotations;
+    absl::flat_hash_map<uint32, absl::string_view> correlation_map;
+  };
+  const uint64 max_size_;
+  AnnotationMapImpl map_;
+
+  TF_DISALLOW_COPY_AND_ASSIGN(AnnotationMap);
+};
+
+class RocmTraceCollector {
+ public:
+  explicit RocmTraceCollector(const RocmTraceCollectorOptions& options)
+      : options_(options), annotation_map_(options.max_annotation_strings) {}
+  virtual ~RocmTraceCollector() {}
+
+  virtual void AddEvent(RocmTracerEvent&& event) = 0;
+  virtual void OnEventsDropped(const std::string& reason,
+                               uint32 num_events) = 0;
+  virtual void Flush() = 0;
+
+  AnnotationMap* annotation_map() { return &annotation_map_; }
+
+ protected:
+  RocmTraceCollectorOptions options_;
+
+ private:
+  AnnotationMap annotation_map_;
+
+  TF_DISALLOW_COPY_AND_ASSIGN(RocmTraceCollector);
+};
+
+// forward declarations for callback functors
+class RocmApiCallbackImpl;
+class RocmActivityCallbackImpl;
+
+// The class use to enable cupti callback/activity API and forward the collected
+// trace events to RocmTraceCollector. There should be only one RocmTracer
+// per process.
+class RocmTracer {
+ public:
+  // Returns a pointer to singleton RocmTracer.
+  static RocmTracer* GetRocmTracerSingleton();
+
+  // Only one profile session can be live in the same time.
+  bool IsAvailable() const;
+
+  void Enable(const RocmTracerOptions& options, RocmTraceCollector* collector);
+  void Disable();
+
+  void ApiCallbackHandler(uint32_t domain, uint32_t cbid, const void* cbdata);
+  void ActivityCallbackHandler(const char* begin, const char* end);
+
+  static uint64 GetTimestamp();
+  static int NumGpus();
+
+  void AddToPendingActivityRecords(uint32 correlation_id) {
+    pending_activity_records_.Add(correlation_id);
+  }
+
+  void RemoveFromPendingActivityRecords(uint32 correlation_id) {
+    pending_activity_records_.Remove(correlation_id);
+  }
+
+  void ClearPendingActivityRecordsCount() { pending_activity_records_.Clear(); }
+
+  size_t GetPendingActivityRecordsCount() {
+    return pending_activity_records_.Count();
+  }
+
+ protected:
+  // protected constructor for injecting mock cupti interface for testing.
+  explicit RocmTracer() : num_gpus_(NumGpus()) {}
+
+ private:
+  Status EnableApiTracing();
+  Status DisableApiTracing();
+
+  Status EnableActivityTracing();
+  Status DisableActivityTracing();
+
+  int num_gpus_;
+  absl::optional<RocmTracerOptions> options_;
+  RocmTraceCollector* collector_ = nullptr;
+
+  bool api_tracing_enabled_ = false;
+  bool activity_tracing_enabled_ = false;
+
+  RocmApiCallbackImpl* api_cb_impl_;
```

Comment:
Could you point me to where api_cb_impl_ and activity_cb_impl_ are being used?

Reply:
https://github.com/tensorflow/tensorflow/pull/47508/files#diff-ed24af3bc5fc7ee922e502256a716aa78ef7ce9cf479a8d56d70d2a147078309R939

https://github.com/tensorflow/tensorflow/pull/47508/files#diff-ed24af3bc5fc7ee922e502256a716aa78ef7ce9cf479a8d56d70d2a147078309R940